src/python/pants/core/goals/init.py

```diff
@@ -0,0 +1,224 @@
+
+
+class PutativeTargets(DeduplicatedCollection[PutativeTarget]):
+    sort_input = True
+
+    @classmethod
+    def merge(cls, tgts_iters: Iterable["PutativeTargets"]) -> "PutativeTargets":
+        all_tgts: List[PutativeTarget] = []
+        for tgts in tgts_iters:
+            all_tgts.extend(tgts)
+        return cls(all_tgts)
+
+
+class InitSubsystem(LineOriented, GoalSubsystem):
+    name = "init"
+    help = "Generate config for working with Pants."
+
+
+class Init(Goal):
+    subsystem_cls = InitSubsystem
+
+
+@dataclass(frozen=True)
+class EditBuildFilesRequest:
+    putative_targets: PutativeTargets
+
+
+@dataclass(frozen=True)
+class EditedBuildFiles:
+    digest: Digest
+    created_paths: Tuple[str, ...]
+    updated_paths: Tuple[str, ...]
+
+
+def group_by_build_file(ptgts: Iterable[PutativeTarget]) -> Dict[str, List[PutativeTarget]]:
+    ret = defaultdict(list)
+    for ptgt in ptgts:
+        ret[ptgt.build_file_path].append(ptgt)
+    return ret
+
+
+@rule
+async def edit_build_files(req: EditBuildFilesRequest) -> EditedBuildFiles:
+    ptgts_by_build_file = group_by_build_file(req.putative_targets)
+    existing_build_files_contents = await Get(DigestContents, PathGlobs(ptgts_by_build_file.keys()))
+    existing_build_files_contents_by_path = {
+        ebfc.path: ebfc.content for ebfc in existing_build_files_contents
+    }
+
+    def make_content(bf_path: str, pts: Iterable[PutativeTarget]) -> FileContent:
+        existing_content = existing_build_files_contents_by_path.get(bf_path)
+        new_content = ([] if existing_content is None else [existing_content.decode()]) + [
+            pt.generate_build_file_stanza() for pt in pts
+        ]
+        new_content_bytes = "\n\n".join(new_content).encode()
+        return FileContent(bf_path, new_content_bytes)
+
+    new_digest = await Get(
+        Digest,
+        CreateDigest([make_content(path, ptgts) for path, ptgts in ptgts_by_build_file.items()]),
+    )
+
+    updated = set(existing_build_files_contents_by_path.keys())
+    created = set(ptgts_by_build_file.keys()) - updated
+    return EditedBuildFiles(new_digest, tuple(sorted(created)), tuple(sorted(updated)))
+
+
+@goal_rule
+async def init(
+    init_subsystem: InitSubsystem,
+    console: Console,
+    workspace: Workspace,
+    union_membership: UnionMembership,
+) -> Init:
+    putative_target_request_types = union_membership[PutativeTargetsRequest]
+    putative_target_reqs = [req_type() for req_type in putative_target_request_types]
+    putative_targets_results = await MultiGet(
+        Get(PutativeTargets, PutativeTargetsRequest, req) for req in putative_target_reqs
+    )
+    all_putative_targets = PutativeTargets.merge(putative_targets_results)
+    all_existing_tgts = await Get(Targets, AddressSpecs([DescendantAddresses("")]))
+    addr_to_existing_tgt = {
+        f"{tgt.address.spec_path}:{tgt.address.target_name}": tgt for tgt in all_existing_tgts
```

Comment:
This should be the same as `tgt.address.spec`, given that you make the change to use `UnexpandedTargets` so that you only have BUILD targets in play rather than File targets.

Reply:
Fixed to use Address directly